assets/shiny-server.js

```diff
@@ -312,10 +431,74 @@
         channel.onmessage({data: payload});
       }
     };
+
+    // Open a new SockJS connection and assign it.
+    this._openConnection = function(){
+      var url = self._sockjsUrl;
+      if (!self._first){
+        // Communicate to the server that we're intending to re-use an existing ID.
+        url = url.replace(/\/n=/, '/o=');
+      }
+      var conn = new SockJS(url,
+        null,{protocols_whitelist: self._whitelist});
+      conn.onopen = self.onConnOpen;
+      conn.onclose = self.onConnClose;
+      conn.onmessage = self.onConnMessage;
+
+      self._conn = conn;
+    };
+
+    // Open a connection now.
+		this._openConnection();
+
+		this._parseMultiplexData = function(msg) {
+			try {
+				var m = /^(\d+)\|(m|o|c|r)\|([\s\S]*)$/m.exec(msg);
+				if (!m)
+					return null;
+				msg = {
+					id: m[1],
+					method: m[2],
+					payload: m[3]
+				};
+
+				switch (msg.method) {
+					case 'm':
+						break;
+					case 'o':
+						if (msg.payload.length === 0)
+							return null;
+						break;
+					case 'c':
+						try {
+							msg.payload = JSON.parse(msg.payload);
+						} catch(e) {
+							return null;
+						}
+						break;
+          case 'r':
+            self._disconnected = true;
+            if (msg.payload.length > 0){
```

Comment:
Yeah. my view was that 'r' was a way for the server to communicate a fatal error to the user in response to an action they tried to take. i.e. I click "reconnect" and the server's unable to fulfill it and now all I can do is "reload". We alert to explain. This could eventually become a UI flash.